File: 1. 단계별로_풀어보기/10~19/16. 스택, 큐, 덱 1/9. 28279.js
// 입력
const fs = require('fs')
const inputArr = fs.readFileSync(0).toString().trim().split('\n')

// 입력 파싱
const N = +inputArr[0]
const queryArr = inputArr.slice(1).map(str => str.split(' ').map(Number))

// 출력 결과 변수
let resultStr = ''

// 큐 역할을 할 배열
const deque = Array.from({length: 2000001}).fill(0)
let headIdx = 1000000
let tailIdx = 999999

// 연산
queryArr.forEach(query => {
  const order = query[0]

  const isEmpty = headIdx > tailIdx

  if (order === 1) {
    const X = query[1]
    deque[--headIdx] = X
  } // ::
  else if (order === 2) {
    const X = query[1]
    deque[++tailIdx] = X
  } // ::
  else if (order === 3) {
    if (isEmpty) {
      resultStr += '-1\n'
    } // ::
    else {
      resultStr += `${deque[headIdx++]}\n`
    }
  } // ::
  else if (order === 4) {
    if (isEmpty) {
      resultStr += '-1\n'
    } // ::
    else {
      resultStr += `${deque[tailIdx--]}\n`
    }
  } // ::
  else if (order === 5) {
    resultStr += `${tailIdx - headIdx + 1}\n`
  } // ::
  else if (order === 6) {
    resultStr += `${isEmpty ? 1 : 0}\n`
  } // ::
  else if (order === 7) {
    if (isEmpty) {
      resultStr += '-1\n'
    } // ::
    else {
      resultStr += `${deque[headIdx]}\n`
    }
  } // :;
  else {
    if (isEmpty) {
      resultStr += '-1\n'
    } // ::
    else {
      resultStr += `${deque[tailIdx]}\n`
    }
  }
})

process.stdout.write(resultStr)

File: 1. 단계별로_풀어보기/10~19/16. 스택, 큐, 덱 1/9. 28279.py
from collections import deque
import sys

input = sys.stdin.readline
output = sys.stdout.write

N = int(input())
queryArr = list(list(map(int, input().split())) for i in range(N))

deque = deque()

for query in queryArr:
  order = query[0]

  match (order):
    case 1:
      X = query[1]
      deque.appendleft(X)
    case 2:
      X = query[1]
      deque.append(X)
    case 3:
      if len(deque) > 0:
        output(f"{deque.popleft()}\n")
      else:
        output(f"-1\n")
    case 4:
      if len(deque) > 0:
        output(f"{deque.pop()}\n")
      else:
        output(f"-1\n")
    case 5:
      output(f"{len(deque)}\n")
    case 6:
      output(f"{1 if len(deque) == 0 else 0}\n")
    case 7:
      if len(deque) > 0:
        output(f"{deque[0]}\n")
      else:
        output(f"-1\n")
    case 8:
      if len(deque) > 0:
        output(f"{deque[-1]}\n")
      else:
        output(f"-1\n")